models/classes/Observer/GCP/PubSubClientFactory.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA.
+ */
+
+declare(strict_types=1);
+
+namespace oat\tao\model\Observer\GCP;
+
+use ErrorException;
+use Google\Cloud\PubSub\PubSubClient;
+
+class PubSubClientFactory
+{
+    /** @var array */
+    private $environmentVars;
+
+    public function __construct(array $environmentVars = null)
+    {
+        $this->environmentVars = $environmentVars ?? $_ENV;
+    }
+
+    /**
+     * @return \Google\Cloud\PubSub\PubSubClient
+     *
+     * @throws ErrorException
+     */
+    public function create(array $config = [])
+    {
+        if (!class_exists('Google\Cloud\PubSub\PubSubClient') || !$this->getPubSubFilePath()) {
```

Comment:
Suggestion: I think you may let PHP itself expand the FQN for the class, benefiting from the `use` clause you already have

```suggestion
        if (!class_exists(PubSubClient::class) || !$this->getPubSubFilePath()) {
```

Reply:
In this case, PHP may fail if `PubSubClient` class does not exist.
I think we should keep it as a string instead of calling `::class`.

Reply:
It should not fail but just check if the class exists, so `::class` works like if we were copypasting the FQN here (I've double-checked just in case :D).

```
$ cat expand-class-fqn.php 
<?php

use Google\Cloud\PubSub\PubSubClient;

use Some\Random\ClassName;

if (!class_exists(PubSubClient::class)) {
        echo PubSubClient::class." does not exist\n";
}
if (!class_exists(ClassName::class)) {
        echo ClassName::class." does not exist\n";
}

$ php expand-class-fqn.php 
Google\Cloud\PubSub\PubSubClient does not exist
Some\Random\ClassName does not exist

```

Reply:
https://github.com/oat-sa/tao-core/pull/3395#discussion_r802604349

Reply:
Done :)